hudi-client/hudi-client-common/src/main/java/org/apache/hudi/client/BaseHoodieTableServiceClient.java

```diff
@@ -298,12 +314,56 @@ protected HoodieWriteMetadata<O> compact(String compactionInstantTime, boolean s
       table.getMetaClient().reloadActiveTimeline();
     }
     compactionTimer = metrics.getCompactionCtx();
+    // start commit in MDT if enabled
+    Option<HoodieTableMetadataWriter> metadataWriterOpt = getMetadataWriterFunc.apply(compactionInstantTime, table.getMetaClient());
+    if (metadataWriterOpt.isPresent()) {
+      metadataWriterOpt.get().reInitWriteClient();
+      metadataWriterOpt.get().startCommit(compactionInstantTime);
+    }
     HoodieWriteMetadata<T> writeMetadata = table.compact(context, compactionInstantTime);
-    HoodieWriteMetadata<O> compactionMetadata = convertToOutputMetadata(writeMetadata);
-    if (shouldComplete && compactionMetadata.getCommitMetadata().isPresent()) {
-      completeCompaction(compactionMetadata.getCommitMetadata().get(), table, compactionInstantTime);
+    HoodieWriteMetadata<T> processedWriteMetadata = writeToMetadata(writeMetadata, compactionInstantTime, metadataWriterOpt);
+
+    HoodieWriteMetadata<O> compactionWriteMetadata = convertToOutputMetadata(processedWriteMetadata);
+    if (shouldComplete) {
+      // don't need to support auto commit flow.
+      // triggering dag for compaction here.
+      commitCompaction(compactionInstantTime, compactionWriteMetadata, Option.of(table), metadataWriterOpt);
     }
-    return compactionMetadata;
+    return compactionWriteMetadata;
+  }
+
+  protected HoodieWriteMetadata<T> writeToMetadata(HoodieWriteMetadata<T> writeMetadata, String compactionInstantTime,
+                                                   Option<HoodieTableMetadataWriter> metadataWriterOpt) {
+    return writeMetadata;
+  }
+
+  protected abstract Pair<List<HoodieWriteStat>, List<HoodieWriteStat>> processAndFetchHoodieWriteStats(HoodieWriteMetadata<O> writeMetadata);
+
+  public void commitCompaction(String compactionInstantTime, HoodieWriteMetadata<O> compactionWriteMetadata, Option<HoodieTable> tableOpt,
+                               Option<HoodieTableMetadataWriter> metadataWriterOpt) {
+    // dereferencing the write dag for compaction for the first time.
```

Comment:
committing compaction. 
here is where (processAndFetchHoodieWriteStats), we trigger the dag which will end up writing to both data table and metadata table.